src/async.rs

```diff
@@ -70,70 +32,120 @@ impl Async {
+    ///         { todo!() }
+    /// }
+    ///
+    /// impl Encoder<Frame> for Codec {
+    ///     type Error = std::io::Error;
+    ///     // encoding frame into buffer
+    ///     fn encode(&mut self, _: Frame, _: &mut BytesMut) -> Result<(), std::io::Error>
+    ///        { todo!() }
+    /// }
+    /// #
     /// # fn main() {
     /// let iface = Iface::new("mytun%d", Mode::Tun).unwrap();
-    /// let name = iface.name().to_owned();
-    /// // Bring the interface up by `ip addr add IP dev $name; ip link set up dev $name`
-    /// let core = Core::new().unwrap();
-    /// let async = Async::new(iface, &core.handle()).unwrap();
-    /// let (sink, stream) = async.split();
+    /// let iface = Async::new(iface).unwrap();
+    /// let (sink, stream) = Codec.framed(iface).split();
     /// # }
     /// ```
-    pub fn new(iface: Iface, handle: &Handle) -> Result<Self> {
+    pub fn new(iface: Iface) -> Result<Self> {
         iface.set_non_blocking()?;
         Ok(Async {
-            mio: PollEvented::new(MioWrapper { iface }, handle)?,
-            recv_bufsize: 1542,
+            inner: AsyncFd::new(iface)?,
         })
     }
-    /// Sets the receive buffer size.
-    ///
-    /// When receiving a packet, a buffer of this size is allocated and the packet read into it.
-    /// This configures the size of the buffer.
-    ///
-    /// This needs to be called when the interface's MTU is changed from the default 1500. The
-    /// default should be enough otherwise.
-    pub fn set_recv_bufsize(&mut self, bufsize: usize) {
-        self.recv_bufsize = bufsize;
+
+    /// Receives a packet from the interface.
+    pub async fn recv(&self, out: &mut [u8]) -> io::Result<usize> {
+        loop {
+            let mut guard = self.inner.readable().await?;
+
+            match guard.try_io(|inner| inner.get_ref().recv(out)) {
+                Ok(result) => return result,
+                Err(_would_block) => continue,
+            }
+        }
+    }
+    /// Sends a packet into the interface.
+    pub async fn send(&mut self, buf: &[u8]) -> Result<usize> {
+        loop {
+            let mut guard = self.inner.writable().await?;
+
+            match guard.try_io(|inner| inner.get_ref().send(buf)) {
+                Ok(result) => return result,
+                Err(_would_block) => continue,
+            }
+        }
     }
 }
 
-impl Stream for Async {
-    type Item = Vec<u8>;
-    type Error = Error;
-    fn poll(&mut self) -> FPoll<Option<Self::Item>, Self::Error> {
-        // TODO Reuse buffer?
-        let mut buffer = vec![0; self.recv_bufsize];
-        match self.mio.read(&mut buffer) {
-            Ok(size) => {
-                buffer.resize(size, 0);
-                Ok(FAsync::Ready(Some(buffer)))
-            },
-            Err(ref e) if e.kind() == ErrorKind::WouldBlock => Ok(FAsync::NotReady),
-            Err(e) => Err(e),
+impl AsyncRead for Async {
```

Comment:
I'm afraid implementing AsyncRead/AsyncWrite for a pcap interface is conceptually wrong. These trains are for things that streams of bytes without being split into messages.

The pcap interface is more message oriented ‒  more like a UdpSocket, where message boundaries have a meaning. Exposing interface like this could lead to weird hard to reproduce bugs where sometimes some kind of buffer is accidentally split into two messages or two messages joined into one buffer.